### PR TITLE
Add `type-check` and unify tsconfig

### PR DESCRIPTION
**Pull request by cb1kenobi**
This adds a new `pnpm type-check` that leverages `turbo` to validate the TypeScript code. This can be run at the top-level or for an individual package.

The `test-lint` workflow will run it after linting and doing the prettier check.

As apart of this effort, each package's `tsconfig.json` has been simplified. There's a new top-level `tsconfig.base.json` file that extends the Vercel Style Guide for TypeScript. Each package's `tsconfig.json` has been audited and previously suppressed rules that no longer apply have been removed. The result is each package's `tsconfig.json` is greatly simplified and we can control common settings in the base config while keeping the flexibility of package-level overrides.

Lastly, in `package/cli`, `pnpm build` calls `scripts/build.mjs` which calls `scripts/compile-templates.mjs`. The `compile-templates.mjs` file was generating invalid TypeScript code. I've fixed it and now it's happier than ever.

Note: In order to run `pnpm type-check`, you must first `pnpm build` because we need the `.d.ts` definition files.

### CODE REVIEW

File: utils/build.mjs
Comment (styfle) [inline utils/build.mjs:62]: Can these be moved to `tsconfig.base.json` instead?

File: tsconfig.base.json
Comment (styfle) [inline tsconfig.base.json:9]: How about adding `noFallthroughCasesInSwitch: true` since most of the others used that too?

File: tsconfig.base.json
Comment (styfle) [inline tsconfig.base.json:4]: How about adding `outDir: "./dist"` since it looks like all of them use it?

File: tsconfig.base.json
Comment (styfle) [inline tsconfig.base.json]:
```suggestion
    "strict": true,
```

Lets enable all strict features like we had before

File: tsconfig.base.json
Comment (styfle) [inline tsconfig.base.json:3]: Let's make sure @Ethan-Arrowood reviews this since there were changes coming to the style guide:

- https://github.com/vercel/style-guide/pull/60

File: utils/build.mjs
Comment (cb1kenobi) [reply utils/build.mjs:62]: No because the `type-check` scripts all run `tsc --noEmit` which cannot be combined with `emitDeclarationOnly`. This means we need to explicitly pass in these flags given the context.

File: tsconfig.base.json
Comment (cb1kenobi) [reply tsconfig.base.json:9]: It's already in the style guide base config: https://github.com/vercel/style-guide/blob/canary/typescript.json#L6. :)

File: tsconfig.base.json
Comment (cb1kenobi) [reply tsconfig.base.json]: It's already in the style guide base config: https://github.com/vercel/style-guide/blob/canary/typescript.json#L9. :)

File: tsconfig.base.json
Comment (cb1kenobi) [reply tsconfig.base.json:4]: TypeScript is funny with relative paths. If I put `outDir` in the base config, it will output all `.js` files into a top-level `dist` directory and it doesn't take long before the build fails because a dependent package built files can't be found. We need to leave `outDir` in each package's `tsconfig.json`.

File: package.json
Comment (Ethan-Arrowood) [inline package.json]: This should probably be a dev dep right? I guess it doesn't matter as its the root package.json

File: packages/cli/tsconfig.json
Comment (Ethan-Arrowood) [inline packages/cli/tsconfig.json:6]: So these are set to `false` because our `cli` code isn't compatible with the strictness of `@vercel/style-guide/typescript`, right? 

We should track this in Linear as a follow up to improve our code quality to match the strictness of the style-guide. Sorta like API

File: packages/node/tsconfig.json
Comment (Ethan-Arrowood) [inline packages/node/tsconfig.json:5]: I prefer `node16` over `nodenext` as its clearer that it won't magically change in the future

File: tsconfig.base.json
Comment (Ethan-Arrowood) [reply tsconfig.base.json:3]: Until that PR is merged and released (in probably a new major) we are fine to stick to the existing config. It is good enough afaik.

File: tsconfig.base.json
Comment (Ethan-Arrowood) [reply tsconfig.base.json:4]: I prefer this is omitted from the base config. 

We could have a `tsconfig.build.json` that assigns this property and others (like `declaration`). This way the `.build.json` can include _just_ `src/` files and then the default config can include _everything_ (i.e. `src/`, `test/`, etc.).

**Review by Ethan-Arrowood — DISMISSED** [review]
Looking good. I don't think I have any blocking changes.

File: tsconfig.base.json
Comment (cb1kenobi) [reply tsconfig.base.json:4]: I didn't want to touch anything having to do with paths (e.g. `include`, `exclude`, `outDir`) because it never works in monorepos. All of those relative paths are being resolved relative to the `tsconfig` file and that's not what we want.

File: tsconfig.base.json
Comment (cb1kenobi) [reply tsconfig.base.json]: Removed `strictNullChecks`.

File: package.json
Comment (cb1kenobi) [reply package.json]: Good catch! Fixed in fb26f5b61c7ea6b0d2886652672a52f74c9b4cea.

File: packages/cli/tsconfig.json
Comment (cb1kenobi) [reply packages/cli/tsconfig.json:6]: Yep.

File: packages/node/tsconfig.json
Comment (cb1kenobi) [reply packages/node/tsconfig.json:5]: I dunno what exactly the differences are between `node16` and `nodenext`. Given I didn't change the value, I'm leaning more towards not changing it in this PR. I'd really like to not break anything, though I doubt it would break anything. @styfle what do you think?

File: tsconfig.base.json
Comment (Ethan-Arrowood) [reply tsconfig.base.json:4]: Another great point

File: packages/node/tsconfig.json
Comment (Ethan-Arrowood) [reply packages/node/tsconfig.json:5]: fair enough. we can make another pass on this in the future

File: utils/build.mjs
Comment (styfle) [reply utils/build.mjs:62]: Ok make sense. And it seems like `--declaration` is here because we want it for builders but we don't want it for cli 👍

**Review by styfle — APPROVED** [review]
Looks good!

In a future PR, it would be nice to set `noUncheckedIndexedAccess: true` to follow the style guide.

**Review by Ethan-Arrowood — APPROVED** [review]

**Review by ecklf — APPROVED** [review]
Approving to unblock compute codeowners requirement.

**Pull request merged**
